### PR TITLE
shell: Add defaultDexConnector option to automatically trigger login on a predefined Dex connector

### PR DESCRIPTION
**Pull request by JBWatenbergScality**
**Component**:

shell

**Context**: 

Dex is providing the possibility to use several connectors to federate OIDC providers. However under some conditions we may need to force the usage of a specific connector.

**Summary**:

This PR add a `defaultDexConnector` option to enable this.

This option can be added to the auth section of micro application `runtime-app-configuration`.
It intercept the `authorization_endpoint` of the oidc web finger and append it the `connector_id` query parameter.

The connector choice can be forced by adding the `displayLoginChoice` query parameter to the current URL.
This enable to maintain UI login possible when the default connector is unavailable.

Additionaly as Dex is not supporting `end_session_endpoint` the logout is reviewed to clean the cookies of the
domain and clearing the localStorage entries of oidc client when `providerLogout` is set to true.

**Acceptance criteria**: 

We shall be able to login directly using the default dex connector when the property is defined

### CODE REVIEW

File: shell-ui/src/auth/AuthProvider.js [hunk 1]
```diff
@@ -39,6 +57,9 @@ function OAuth2AuthProvider({ children }: { children: Node }) {
     loadUserInfo: true,
     automaticSilentRenew: true,
     monitorSession: false,
+    MetadataServiceCtor: authConfig.defaultDexConnector
```
Comment (ChengYanJin) [inline shell-ui/src/auth/AuthProvider.js:60]: I am wondering should we add this `defaultDexConnector` prop to type `OIDCConfig`?

File: shell-ui/src/auth/AuthProvider.js [hunk 2]
```diff
@@ -39,6 +57,9 @@ function OAuth2AuthProvider({ children }: { children: Node }) {
...
+      ? defaultDexConnectorMetadataService(authConfig.defaultDexConnector)
```
Comment (ChengYanJin) [inline shell-ui/src/auth/AuthProvider.js:61]: It's more a question: 
If we want to use Dex to authenticate the users against a connector-based interface, how should we set the `defaultDexConnector` in `runtime-app-configuration`?

File: shell-ui/src/auth/AuthProvider.js [hunk 3]
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
...
+            if (e.message === 'no end session endpoint') {
+              console.log(
+                "Oidc provider doesn't support end session endpoint, fallback to clearing document cookies",
+              );
+              document.cookie.split(';').forEach(function (c) {
+                document.cookie =
+                  c.trim().split('=')[0] +
+                  '=;' +
+                  'expires=Thu, 01 Jan 1970 00:00:00 UTC;';
+              });
```
Comment (gdemonet) [inline shell-ui/src/auth/AuthProvider.js:146]: What's that? We split the cookie and assign a new value for each? How does it work? Is it that assigning a string to `document.cookie` is actually interpreted as an instruction, rather than describing the whole "cookies state"?

File: shell-ui/src/auth/AuthProvider.js [hunk 4]
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
...
+      if (authConfig.kind === 'OAuth2Proxy') {
+        throw new Error('OAuth2Proxy authentication kind is not yet supported');
+      }
+      if (auth && auth.userManager) {
+        auth.userManager.revokeAccessToken();
+        if (authConfig.providerLogout) {
+          auth.userManager.signoutRedirect().catch((e) => {
+            if (e.message === 'no end session endpoint') {
+              console.log(
+                "Oidc provider doesn't support end session endpoint, fallback to clearing document cookies",
```
Comment (gdemonet) [inline shell-ui/src/auth/AuthProvider.js]:
```suggestion
                "OIDC provider doesn't support end session endpoint, fallback to clearing document cookies",
```

File: shell-ui/src/auth/AuthProvider.js [hunk 5]
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
...
+                  c.trim().split('=')[0] +
+                  '=;' +
+                  'expires=Thu, 01 Jan 1970 00:00:00 UTC;';
+              });
+            } else {
+              console.error(e);
+            }
+            auth.userManager.removeUser();
+            location.reload();
+          });
```
Comment (gdemonet) [inline shell-ui/src/auth/AuthProvider.js]: This whole `.catch` block doesn't seem related to the change in this PR (which is about a pre-defined Dex connector). Would be nice to at least use a different commit :wink:

File: shell-ui/src/auth/AuthProvider.js (hunk 2 above)
Comment (gdemonet) [reply shell-ui/src/auth/AuthProvider.js:61]: Didn't we start documenting all this rework of shell-ui? Would be useful IMO, both now for reviewers catching up with the progress, and in the future for maintenance / onboarding.

**Review by gdemonet — CHANGES_REQUESTED** [review]
No need to tackle documentation as part of this PR :wink:

File: shell-ui/src/auth/AuthProvider.js (hunk 3 above)
Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js:146]: Yes using `document.cookie` we can only set or update a single cookie at a time (https://developer.mozilla.org/en-US/docs/web/api/document/cookie#write_a_new_cookie). What I'm trying to achieve here is a hack to logout the users from the federated OIDC Provider behind Dex because Dex is not supporting the end session endpoint and then the session on the original OIDC provider can't be closed in any other way 😕 .

File: shell-ui/src/auth/AuthProvider.js (hunk 1 above)
Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js:60]: Done in db1812c1d0d9a5d1e21a20caefd003b31b9a1a75

File: shell-ui/src/auth/AuthProvider.js (hunk 2 above)
Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js:61]: `defaultDexConnector` should take the connector id as defined in dex configuration.
For example given the following Dex connectors configuration :
```
       connectors:
            - type: oidc
              id: keycloak
              name: Keycloak
              config:
                   //....
```

Providing `keycloak` as `defaultDexConnector` in `runtime-app-configuration` will automatically redirect the user to this connector.

Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js:61]: I've opened #3476 regarding the documentation and will address it as soon as possible

File: shell-ui/src/auth/AuthProvider.js (hunk 5 above)
Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js]: Done in e6566809c79b0f9ea3cb11fa81d856096bb04d06

File: shell-ui/src/auth/AuthProvider.js (hunk 4 above)
Comment (JBWatenbergScality) [reply shell-ui/src/auth/AuthProvider.js]: Done in e6566809c79b0f9ea3cb11fa81d856096bb04d06

**Review by gdemonet — APPROVED** [review]
